File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
peopleN = int(input())
grid = []

for i in range(peopleN):
  grid.append(list(map(int, input().split())))

peopleArr = []

for i in range(peopleN):
  peopleArr.append(i)

answer = float('inf')

def dfs(arr, start):
  global answer

  if len(arr) == peopleN/2:
    arr2 = list(filter(lambda x : x not in arr, peopleArr))
    team1 = 0
    team2 = 0
    for i in range(len(arr)-1):
      for j in range(i+1, len(arr)):
        team1 += grid[arr[i]][arr[j]] + grid[arr[j]][arr[i]]

    for i in range(len(arr)-1):
      for j in range(i+1, len(arr)):
        team2 += grid[arr2[i]][arr2[j]] + grid[arr2[j]][arr2[i]]

    answer = min(abs(team1 - team2), answer)
    return

  for i in range(start, peopleN):
    arr.append(i)
    dfs(arr, i+1)
    arr.pop()

dfs([], 0)

print(answer)

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// 첫번째줄에 n이 주어진다
// 두번재줄부터 그리드가 주어지고 각 능력치가 주어진다
// N은 항상 짝수고 반반씩 나눠서 능력치의 차이를 최소로 팀으로 나눈다
// 그렇다면 백트레킹으로 n/2 만큼 조합을 먼저 만들고나서
// 각 케이스마다 팀마다 능력치의 합을 구해야된다
const solution = (input) => {
  const n = Number(input[0]);
  const grid = [];
  const allTeam = Array(n)
    .fill(0)
    .map((v, index) => index);

  for (let i = 1; i <= n; i++) {
    grid.push(input[i].split(" ").map((v) => Number(v)));
  }

  let answer = Infinity;

  const backTrack = (arr, start) => {
    if (arr.length === n / 2) {
      const firstTeam = [...arr];
      const secondTeam = allTeam.filter((v) => !firstTeam.includes(v));
      let firstTeamScore = 0;
      let secondTeamScore = 0;

      for (let i = 0; i < n / 2; i++) {
        for (let j = 0; j < n / 2; j++) {
          const f1 = firstTeam[i];
          const f2 = firstTeam[j];
          const s1 = secondTeam[i];
          const s2 = secondTeam[j];
          firstTeamScore += grid[f1][f2];
          secondTeamScore += grid[s1][s2];
        }
      }

      answer = Math.min(answer, Math.abs(firstTeamScore - secondTeamScore));
      return;
    }

    for (let i = start; i < n; i++) {
      arr.push(i);
      backTrack(arr, i + 1);
      arr.pop();
    }
  };

  backTrack([], 0);

  console.log(answer);
};

solution(input);
